File: core/src/main/java/com/starshootercity/abilities/AquaAffinity.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import com.starshootercity.OriginsReborn;
import net.kyori.adventure.key.Key;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffectType;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class AquaAffinity implements VisibleAbility, BreakSpeedModifierAbility {
    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:aqua_affinity");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor("Вы можете ломать блоки под водой так же, как и на суше.", OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor("Родство с водой", OriginSwapper.LineData.LineComponent.LineType.TITLE);
    }

    @Override
    public BlockMiningContext provideContextFor(Player player) {
        return new BlockMiningContext(
                player.getInventory().getItemInMainHand(),
                player.getPotionEffect(OriginsReborn.getNMSInvoker().getHasteEffect()),
                player.getPotionEffect(OriginsReborn.getNMSInvoker().getMiningFatigueEffect()),
                player.getPotionEffect(PotionEffectType.CONDUIT_POWER),
                true,
                true,
                true
        );
    }

    @Override
    public boolean shouldActivate(Player player) {
        return player.isInWater();
    }
}

File: core/src/main/java/com/starshootercity/abilities/AquaAffinity.kt
package com.starshootercity.abilities

import com.starshootercity.OriginSwapper.LineData.Companion.makeLineFor
import com.starshootercity.OriginSwapper.LineData.LineComponent
import com.starshootercity.OriginsReborn.Companion.NMSInvoker
import com.starshootercity.abilities.BreakSpeedModifierAbility.BlockMiningContext
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.potion.PotionEffectType

class AquaAffinity : VisibleAbility, BreakSpeedModifierAbility {
    override fun getKey(): Key {
        return Key.key("origins:aqua_affinity")
    }

    override fun getDescription(): MutableList<LineComponent?> {
        return makeLineFor("You may break blocks underwater as others do on land.", LineComponent.LineType.DESCRIPTION)
    }

    override fun getTitle(): MutableList<LineComponent?> {
        return makeLineFor("Aqua Affinity", LineComponent.LineType.TITLE)
    }

    override fun provideContextFor(player: Player) = with(player) {
        BlockMiningContext(
            inventory.itemInMainHand,
            getPotionEffect(NMSInvoker.getHasteEffect()),
            getPotionEffect(NMSInvoker.getMiningFatigueEffect()),
            getPotionEffect(PotionEffectType.CONDUIT_POWER),
            true,
            true,
            true
        )
    }

    override fun shouldActivate(player: Player): Boolean {
        return player.isInWater
    }
}
